src/dtree.ml

```diff
@@ -0,0 +1,681 @@
+
+  (** [choose m] returns the index of column to switch on. *)
+  let choose : t -> (int * float) option = fun m ->
+    let kept = discard_cons_free m.clauses in
+    if kept = [||] then None else
+    let sel_partial, score = pick_best_among m kept in
+    let cind = kept.(sel_partial) in
+    Some(cind, score)
+
+  (** [is_exhausted r] returns whether [r] can be applied or not. *)
+  let is_exhausted : clause -> bool = fun { lhs ; nonlin ; freevars ; _ } ->
+    let nonl lhs =
+      (* Verify that there are no non linearity constraints in the remaining
+         terms.  We must check that there are no constraints in the remaining
+         terms and no constraints left partially instantiated. *)
+      let slots = Array.to_list lhs
+                 |> List.filter_map
+                      (function Patt(io, _, _) -> io | _ -> None) in
+      let slots_uniq = List.sort_uniq Int.compare slots in
+      List.same_length slots slots_uniq
+      && not @@ List.exists (fun s -> NlScorable.constrained s nonlin)
+          slots_uniq in
+    let ripe lhs =
+      (* [ripe l] returns whether [lhs] can be applied. *)
+      Array.for_all (function Patt(_, _, [||]) -> true | _ -> false) lhs
+      && nonl lhs in
+    NlScorable.is_empty nonlin && FvScorable.is_empty freevars
+    && (lhs = [||] || ripe lhs)
+
+  (** [yield m] yields a clause to be applied. *)
+  let yield : t -> decision = fun ({ clauses ; _ } as m) ->
+    try
+      if !ordered_rules
+      then let fc = List.hd clauses in
+        if is_exhausted fc then Yield(fc) else raise Not_found
+      else let r = List.find is_exhausted clauses in
+      Yield(r)
+    with Not_found ->
+      let nlcstrs = List.map (fun r -> r.nonlin) m.clauses in
+      let rnl = match NlScorable.choose nlcstrs with
+        | Some(c, i) -> (Nl(c), i)
+        | None       -> (Unavailable, 0.) in
+      let fvcstrs = List.map (fun r -> r.freevars) m.clauses in
+      let rfv = match FvScorable.choose fvcstrs with
+        | Some(c, i) -> (Fv(c), i)
+        | None       -> (Unavailable, 0.) in
+      let rs = match choose m with
+        | None       -> (Unavailable, 0.)
+        | Some(c, i) -> (Sp(c), i) in
+      let r = [| rnl ; rfv ; rs |] in
+      let best =
+        if Array.for_all (fun (x, _) -> x = Unavailable) r then
+          (Unavailable, 0.) else
+          Array.max ~cmp:(fun (_, x) (_, y) -> Float.compare x y) r in
+      match fst best with
+      | Nl(c)          -> NlConstrain(c)
+      | Fv(c)          -> FvConstrain(c)
+      | Sp(c)          -> Specialise(c)
+      | Unavailable    -> Specialise(0)
+
+  (** [get_cons l] extracts, sorts and uniqify terms that are tree
+      constructors in [l].  The actual tree constructor (of type
+      {!type:treecons}) is returned along the original term. *)
+  let get_cons : term list -> (TC.treecons * term) list = fun telst ->
+    let keep_treecons e =
+      if is_treecons e then Some(treecons_of_term e, e) else None in
```

Comment:
It seems to be the only place where treecons_of_term is used. So, why not defining it so that we do not need to call is_treecons before, that is, define keep_treecons as treecons_of_term and adapt the output.